crates/bevy_reflect/src/generics.rs

```diff
@@ -0,0 +1,338 @@
+    ///
+    /// An example would be `N` in `struct Foo<const N: usize>`.
+    Const(ConstParamInfo),
+}
+
+impl GenericInfo {
+    /// The name of the generic parameter.
+    pub fn name(&self) -> &Cow<'static, str> {
+        match self {
+            Self::Type(info) => info.name(),
+            Self::Const(info) => info.name(),
+        }
+    }
+
+    /// Whether the generic parameter is a const parameter.
+    pub fn is_const(&self) -> bool {
+        match self {
+            Self::Type(_) => false,
+            Self::Const(_) => true,
+        }
+    }
+
+    impl_type_methods!(self => {
+        match self {
+            Self::Type(info) => info.ty(),
+            Self::Const(info) => info.ty(),
+        }
+    });
+}
+
+impl From<TypeParamInfo> for GenericInfo {
+    fn from(info: TypeParamInfo) -> Self {
+        Self::Type(info)
+    }
+}
+
+impl From<ConstParamInfo> for GenericInfo {
+    fn from(info: ConstParamInfo) -> Self {
+        Self::Const(info)
+    }
+}
+
+/// Type information for a generic type parameter.
+///
+/// An example of a type parameter would be `T` in `struct Foo<T>`.
+#[derive(Clone, Debug)]
+pub struct TypeParamInfo {
+    name: Cow<'static, str>,
+    ty: Type,
+    default: Option<Type>,
+}
+
+impl TypeParamInfo {
+    /// Creates a new type parameter with the given name.
+    pub fn new<T: TypePath + ?Sized>(name: impl Into<Cow<'static, str>>) -> Self {
+        Self {
+            name: name.into(),
+            ty: Type::of::<T>(),
+            default: None,
+        }
+    }
+
+    /// Sets the default type for the parameter.
+    pub fn with_default<T: TypePath + ?Sized>(mut self) -> Self {
+        self.default = Some(Type::of::<T>());
+        self
+    }
+
+    /// The name of the type parameter.
+    pub fn name(&self) -> &Cow<'static, str> {
+        &self.name
+    }
+
+    /// The default type for the parameter, if any.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use bevy_reflect::{GenericInfo, Reflect, Typed};
+    /// #[derive(Reflect)]
+    /// struct Foo<T = f32>(T);
+    ///
+    /// let generics = Foo::<String>::type_info().generics();
+    /// let GenericInfo::Type(info) = generics.get_named("T").unwrap() else {
+    ///     panic!("expected a type parameter");
+    /// };
+    ///
+    /// let default = info.default().unwrap();
+    ///
+    /// assert!(default.is::<f32>());
+    /// ```
+    pub fn default(&self) -> Option<&Type> {
+        self.default.as_ref()
+    }
+
+    impl_type_methods!(ty);
+}
+
+/// Type information for a const generic parameter.
+///
+/// An example of a const parameter would be `N` in `struct Foo<const N: usize>`.
+#[derive(Clone, Debug)]
+pub struct ConstParamInfo {
+    name: Cow<'static, str>,
+    ty: Type,
+    // Rust currently only allows certain primitive types in const generic position,
+    // meaning that `Reflect` is guaranteed to be implemented for the default value.
+    default: Option<Arc<dyn Reflect>>,
```

Comment:
What's the reason for an `Arc` instead of a `Box`?

Reply:
This allows us to `Clone` the value, which is required by `TypeInfo`. Alternatively, we could store a `fn() -> Box<dyn Reflect>`.

Reply:
Makes sense. I prefer the `Arc<dyn Reflect>`, but I don't really have strong feelings either way.